idx/metric_definition.go

```diff
@@ -0,0 +1,552 @@
+			internError.Inc()
+			return fmt.Errorf("idx: Failed to intern word in MetricName: %v, %v", node, err)
+		}
+		md.Name.nodes[i] = nodePtr
+	}
+
+	return nil
+}
+
+// SetTags takes a []string which should contain Key/Value pairs
+// in the form of 'key=value'. It splits up the Key and Value for each
+// item, interns them in the object store, and creates a TagKeyValue
+// for them. It then stores all of these in Tags.
+//
+// The items in the input argument should not contain ';'. Each item
+// is a separate Key/Value pair. Do not combine multiple Key/Value pairs
+// into a single index in the []string.
+func (md *MetricDefinition) SetTags(tags []string) {
+	md.Tags.KeyValues = make([]TagKeyValue, len(tags))
+	sort.Strings(tags)
+	for i, tag := range tags {
+		if tag == "=" || tag == "" {
+			log.Error("idx: SetTags: Empty tag, ignoring: ", tag)
+			invalidTag.Inc()
+			continue
+		}
+		if strings.Contains(tag, ";") {
+			log.Errorf("idx: SetTags: Tag %q has an invalid format, ignoring", tag)
+			invalidTag.Inc()
+			continue
+		}
+		eqPos := strings.Index(tag, "=")
+		if eqPos < 0 {
+			log.Errorf("idx: SetTags: Tag %q has an invalid format, ignoring", tag)
+			invalidTag.Inc()
+			continue
+		}
+		key, err := IdxIntern.AddOrGet([]byte(tag[:eqPos]), true)
+		if err != nil {
+			log.Errorf("idx: SetTags: Failed to intern tag %q, %v", tag, err)
+			internError.Inc()
+			continue
+		} else {
+			md.Tags.KeyValues[i].Key = key
+		}
+
+		value, err := IdxIntern.AddOrGet([]byte(tag[eqPos+1:]), true)
+		if err != nil {
+			log.Errorf("idx: SetTags: Failed to intern tag %q, %v", tag, err)
+			internError.Inc()
+			continue
+		} else {
+			md.Tags.KeyValues[i].Value = value
+		}
+	}
+}
+
+// SetId creates and sets the MKey which identifies a metric
+func (md *MetricDefinition) SetId() {
+	sort.Sort(md.Tags.KeyValues)
+	buffer := bytes.NewBufferString(md.Name.String())
+	buffer.WriteByte(0)
+	buffer.WriteString(md.Unit)
+	buffer.WriteByte(0)
+	buffer.WriteString(md.Mtype())
+	buffer.WriteByte(0)
+	fmt.Fprintf(buffer, "%d", md.Interval)
+
+	for _, t := range md.Tags.KeyValues {
+		key, err := IdxIntern.GetStringFromPtr(t.Key)
+		if err != nil {
+			log.Error("idx: Failed to retrieve interned tag key: ", err)
+			internError.Inc()
+			continue
+		}
+		if key == "name" {
+			continue
+		}
+
+		buffer.WriteByte(0)
+		buffer.WriteString(t.String())
+	}
+
+	md.Id = schema.MKey{
+		Key: md5.Sum(buffer.Bytes()),
+		Org: uint32(md.OrgId),
+	}
+}
+
+// ConvertToSchemaMd converts an idx.MetricDefinition to a schema.MetricDefinition
+func (md *MetricDefinition) ConvertToSchemaMd() schema.MetricDefinition {
+	smd := schema.MetricDefinition{
+		Id:         md.Id,
+		OrgId:      md.OrgId,
+		Name:       md.Name.String(),
+		Interval:   md.Interval,
+		Unit:       md.Unit,
+		Mtype:      md.Mtype(),
+		Tags:       md.Tags.Strings(),
+		LastUpdate: md.LastUpdate,
+		Partition:  md.Partition,
+	}
+	smd.NameWithTags() //ensure nameWithTags is set
```

Comment:
why does nameWithTags have to be set? isn't that wasting memory in cases where we don't need the name with tags string?

Reply:
It's only called in `mt-index-cat`, I haven't looked through it enough to determine if we should keep or leave this, so I'm going to leave it in for now.